Detectors/Vertexing/include/DetectorsVertexing/FwdDCAFitterN.h

```diff
@@ -0,0 +1,1266 @@
+  std::array<bool, MAXHYP> mTrPropDone;      // Flag that the tracks are fully propagated to PCA
+  MatSym3D mWeightInv;                       // inverse weight of single track, [sum{M^T E M}]^-1 in EQ.T
+  std::array<int, MAXHYP> mOrder{0};
+  int mCurHyp = 0;
+  int mCrossIDCur = 0;
+  int mCrossIDAlt = -1;
+  bool mAllowAltPreference = true;  // if the fit converges to alternative PCA seed, abandon the current one
+  bool mUseAbsDCA = false;          // use abs. distance minimization rather than chi2
+  bool mPropagateToPCA = true;      // create tracks version propagated to PCA
+  int mMaxIter = 60;                // max number of iterations
+  float mBz = 0;                    // bz field, to be set by user
+  float mMaxR2 = 200. * 200.;       // reject PCA's above this radius
+  float mMaxDXIni = 4.;             // reject (if>0) PCA candidate if tracks DZ exceeds threshold
+  float mMinParamChange = 1e-5;     // stop iterations if largest change of any X is smaller than this
+  float mMinRelChi2Change = 0.98;   // stop iterations is chi2/chi2old > this
+  float mMaxChi2 = 100;             // abs cut on chi2 or abs distance
+  float mMaxDist2ToMergeSeeds = 1.; // merge 2 seeds to their average if their distance^2 is below the threshold
+
+  ClassDefNV(FwdDCAFitterN, 1);
+};
+
+///_________________________________________________________________________
+template <int N, typename... Args>
+template <class... Tr>
+int FwdDCAFitterN<N, Args...>::process(const Tr&... args)
+{
+
+  static_assert(sizeof...(args) == N, "incorrect number of input tracks");
+  assign(0, args...);
+  clear();
+
+  for (int i = 0; i < N; i++) {
+    mTrAux[i].set(*mOrigTrPtr[i], mBz);
+  }
+
+  if (!mCrossings.set(mTrAux[0], *mOrigTrPtr[0], mTrAux[1], *mOrigTrPtr[1])) { // even for N>2 it should be enough to test just 1 loop
+    return 0;                                                                  // no crossing
+  }
+
+  if (mCrossings.nDCA == MAXHYP) { // if there are 2 candidates
+    auto dst2 = (mCrossings.xDCA[0] - mCrossings.xDCA[1]) * (mCrossings.xDCA[0] - mCrossings.xDCA[1]) +
+                (mCrossings.yDCA[0] - mCrossings.yDCA[1]) * (mCrossings.yDCA[0] - mCrossings.yDCA[1]);
+
+    if (dst2 < mMaxDist2ToMergeSeeds) {
+      mCrossings.nDCA = 1;
+      mCrossings.xDCA[0] = 0.5 * (mCrossings.xDCA[0] + mCrossings.xDCA[1]);
+      mCrossings.yDCA[0] = 0.5 * (mCrossings.yDCA[0] + mCrossings.yDCA[1]);
+    }
+  }
+
+  // check all crossings
+  for (int ic = 0; ic < mCrossings.nDCA; ic++) {
+    // check if radius is acceptable
+    if (mCrossings.xDCA[ic] * mCrossings.xDCA[ic] + mCrossings.yDCA[ic] * mCrossings.yDCA[ic] > mMaxR2) {
+      continue;
+    }
+
+    mCrossIDCur = ic;
+    mCrossIDAlt = (mCrossings.nDCA == 2 && mAllowAltPreference) ? 1 - ic : -1; // works for max 2 crossings
+    mNIters[mCurHyp] = 0;
+    mTrPropDone[mCurHyp] = false;
+    mChi2[mCurHyp] = -1.;
+
+    findZatXY_mid(mCurHyp);
```

Comment:
Hi @sadekr OK, just noticed an independent seeding in `findZatXY_mid` (did not check details), but then why helical crossing and the associated rejections are used? The helical estimate with just a few cm level arm is simply not reliable, it is risky to take any decisions based on it.